tests/repo/shallow.c

```diff
@@ -37,3 +39,90 @@ void test_repo_shallow__clears_errors(void)
 	cl_assert_equal_i(0, git_repository_is_shallow(g_repo));
 	cl_assert_equal_p(NULL, git_error_last());
 }
+
+void test_repo_shallow__shallow_oids(void)
+{
+	git_oidarray oids;
+	g_repo = cl_git_sandbox_init("shallow.git");
+
+	cl_git_pass(git_repository_shallow_roots(&oids, g_repo));
+	cl_assert_equal_i(1, oids.count);
+	cl_assert_equal_oid(&g_shallow_oid, &oids.ids[0]);
+
+	git_oidarray_free(&oids);
+}
+
+void test_repo_shallow__cache_clearing(void)
+{
+	git_oidarray oids;
+	git_oid tmp_oid;
+
+	cl_git_pass(git_oid_fromstr(&tmp_oid, "0000000000000000000000000000000000000000"));
+	g_repo = cl_git_sandbox_init("shallow.git");
+
+	cl_git_pass(git_repository_shallow_roots(&oids, g_repo));
+	cl_assert_equal_i(1, oids.count);
+	cl_assert_equal_oid(&g_shallow_oid, &oids.ids[0]);
+	git_oidarray_free(&oids);
+
+	cl_git_mkfile("shallow.git/shallow",
+		"be3563ae3f795b2b4353bcce3a527ad0a4f7f644\n"
+		"0000000000000000000000000000000000000000\n"
+	);
+
+	cl_git_pass(git_repository_shallow_roots(&oids, g_repo));
+	cl_assert_equal_i(2, oids.count);
+	cl_assert((git_oid_equal(&g_shallow_oid, &oids.ids[0]) &&
+		   git_oid_equal(&tmp_oid, &oids.ids[1])) ||
+		  (git_oid_equal(&g_shallow_oid, &oids.ids[1]) &&
+		   git_oid_equal(&tmp_oid, &oids.ids[0])));
+	git_oidarray_free(&oids);
```

Comment:
… so ideally, this testcase should just verify that unlinks are also picked up by dropping all shallow roots. I forgot 😉.